File: dp_labs_and_solutions/dp_labs_and_solutions/dp_java_labs_and_solutions/abstract_factory/solution/Garden.java
// Garden.java

public abstract class Garden
{
	protected Garden(Lettuce aLettucePrototype)
	{
		setLettucePrototype(aLettucePrototype);
	}

	protected void setLettucePrototype(Lettuce aLettucePrototype)
	{
		if (aLettucePrototype == null)
		{
			throw new DpException("void setLettucePrototype(Lettuce aLettucePrototype) -- aLettucePrototype == null.");
		}

		myLettucePrototype = aLettucePrototype.clone();
	}

	public final Lettuce makeLettuce()
	{
		return myLettucePrototype.clone();
	}

	public abstract Tomato makeTomato();

	private Lettuce myLettucePrototype;
}

File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_cpp_labs_and_solutions/abstract_factory/solution/garden.cpp
// garden.cpp

#include "dp_exception.h"
#include "lettuce.h"
#include "garden.h"
#include "tomato.h"

garden::garden(lettuce * a_lettuce_prototype)
{
	set_lettuce_prototype(a_lettuce_prototype);
}

void garden::set_lettuce_prototype(lettuce * a_lettuce_prototype)
{
	if (a_lettuce_prototype == nullptr)
	{
		throw dp_exception("void garden::set_lettuce_prototype(lettuce * a_lettuce_prototype) -- a_lettuce_prototype == nullptr.");
	}

	my_lettuce_prototype = a_lettuce_prototype->clone();
}

lettuce * garden::make_lettuce() const
{
	return my_lettuce_prototype->clone();
}
